core/client/meetinthemiddle_test.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"fmt"
+	"testing"
+)
+
+func TestAllPairsOverlap(t *testing.T) {
+	max := uint64(100)
+	for created := uint64(1); created < max; created++ {
+		for current := created; current < max; current++ {
+			for diff := uint64(0); diff < 1; diff++ {
```

Comment:
It seems that only `diff = 0` is examined. Did you intend trying multiple values?

Reply:
Yes. I'm having a hard time clearly articulating what values of diff are permissible. Maybe you can help me think through this.  eg. I get failure when 
`created: 0,  current: 3, diff: 1` with `newer: [1,3], and older: [4]` no overlap.

Reply:
I think the reason might be that `newer` doesn't go beyond `current`. But it has to, in order to reach to number 4?

Reply:
tbh, I'd drop the `diff` parameter here. The ending of the `Newer` range should be >= of the beginning of `Older`, otherwise they don't always meet in the middle. Effectively, `diff == 0` here is necessary for it to always work.

Reply:
Or you could actually flip it around, and make `diff` "negative". This way it would also work.

Reply:
In many cases, it is ok for the value of `current` as seen by the older algorithm to be ~2x the value of `current` as seen by the newer algorithm.  Maybe this is just `2^(floor(log_2(current))) * 2 - 1 ` which for the above case would be be 3

Reply:
Suppose the second party's `current` is a power of 2, say 1024. The `prev` call will get them straight to 0, so `OlderRevisionsToVerify` will return only [1024]. The only way the caller of `NewerRevisionsToVerify` intersects with it is by providing `current` >= 1024, which is effectively "diff <= 0". It's true in a general sense.

Reply:
I.e. I'm saying "works as intended". Sometimes they indeed can intersect earlier, but the edge case above says the intersection is guaranteed only if the first person monitors at least to the second person's `current`.